bitbybit/src/bitenum.rs

```diff
@@ -1,303 +1,302 @@
-    let enum_vis = input.vis;
-    let enum_attrs = input.attrs;
-
-    let variants = match input.data {
-        Data::Enum(enum_data) => enum_data.variants,
-        _ => panic!("bitenum!: Must be used on enum"),
-    };
-    let mut uses_conditional = false;
-    let emitted_variants: Vec<(&Expr, u128, &Ident, Vec<Attribute>)> = variants.iter().map(|variant| {
-        let variant_name = &variant.ident;
-        let discriminant = variant.discriminant.as_ref().unwrap_or_else(|| panic!("bitenum!: Variant '{}' needs to have a value", variant_name));
-        // Discriminant.0 is the equals sign. 1 is the value
-        let value = &discriminant.1;
-        let string_value = value.to_token_stream().to_string().replace('_', "");
+#[derive(Default)]
+pub(crate) struct Config {
+    explicit_bits: Option<Bits>,
+    explicit_exhaustive: Option<Exhaustive>,
+}
+struct FullConfig {
+    bits: Bits,
+    exhaustive: Exhaustive,
+}
 
-        // Determine the integer value itself. While we don't need it further down (for now),
-        // this ensures that only constants are being used; due to the way how new_with_raw_value()
-        // is written, some expressions would cause compilation issues (e.g. those that refer to other
-        // enum values).
-        let int_value = if let Some(stripped) = string_value.strip_prefix("0x") {
-            u128::from_str_radix(stripped, 16)
-        } else if let Some(stripped) = string_value.strip_prefix("0b") {
-            u128::from_str_radix(stripped, 2)
-        } else if let Some(stripped) = string_value.strip_prefix("0o") {
-            u128::from_str_radix(stripped, 8)
+impl Config {
+    pub(crate) fn parse(&mut self, meta: ParseNestedMeta) -> syn::Result<()> {
+        if meta.path.is_ident("exhaustive") {
+            let result1 = meta.input.parse::<Token![:]>();
+            let result2 = meta.input.parse::<Token![=]>();
+            if result1.is_err() && result2.is_err() {
+                return Err(meta.error(Error::InvalidExhaustiveNextToken));
+            }
+            self.explicit_exhaustive = Some(meta.input.parse()?);
         } else {
-            string_value.parse::<u128>()
-        }.unwrap_or_else(|_| panic!("bitenum!: Error parsing '{}' as integer. Supported: hexadecimal, octal, binary and decimal unsigned integers, but not expressions", string_value));
-
-        if int_value >= (1u128 << bit_count) {
-            panic!("bitenum!: Value {} exceeds the given number of bits", variant_name);
+            let err = || meta.error(Error::InvalidAttribute);
+            let last_segment = meta.path.segments.last().ok_or_else(err)?;
+            let value = last_segment.ident.to_string();
+            let ("u", size) = value.split_at(1) else {
+                return Err(err());
+            };
+            self.explicit_bits = Some(Bits {
+                path: meta.path.clone(),
+                size: size.parse().map_err(|_| err())?,
+            });
         }
```

Comment:
This needs to return an error on size > 64, otherwise we just panic for u128 and bigger before returning an error due to the shift left later.

Reply:
It's already the case. I'm getting the following error:

```
error: The specified storage size is invalid, valid is in range 1..=64
  --> bitbybit-tests/src/bitenum_tests.rs:26:15
   |
26 |     #[bitenum(u67)]
   |               ^^^
```